Assets/Scripts/wallSystem/PlayerController.cs

```diff
@@ -97,104 +98,114 @@ public void ExternalStart(float pickX, float pickY, bool maze = false)
-			Destroy (other.gameObject);
-			if (--localQuota > 0) return;
-			E.Get().CurrTrial.Notify();
-
-			_playingSound = true;
-			E.LogData(
-				TrialProgress.GetCurrTrial().TrialProgress, 
-				TrialProgress.GetCurrTrial().GetRunningTime(),
-				transform,
-				1
-			);
-
-		}
-
-
-		private void ComputeMovement()
-		{
-			//This calculates the current amount of rotation frame rate independent
-			var rotation = Input.GetAxis("Horizontal") * DS.GetData().CharacterData.RotationSpeed * Time.deltaTime;
-
-
-			//This calculates the forward speed frame rate independent
-			_moveDirection = new Vector3(0, 0, Input.GetAxis("Vertical"));
-			_moveDirection = transform.TransformDirection(_moveDirection);
-			_moveDirection *= DS.GetData().CharacterData.MovementSpeed;
-
-			//Here is the movement system
-			const double tolerance = 0.0001;
-
-			//we move iff rotation is 0
-			if (Math.Abs(Mathf.Abs(rotation)) < tolerance)
-				_controller.Move(_moveDirection * Time.deltaTime);
-
-			transform.Rotate(0, rotation, 0);
-
-		}
-	
-		
-		private void Update()
-		{
-	
-			 //This first block is for the initial rotation of the character
-			if (_currDelay < _waitTime)
-			{
-				var angle = 360f * _currDelay / _waitTime + _iniRotation - transform.rotation.eulerAngles.y;
-				transform.Rotate(new Vector3(0, angle, 0));
-			}
-			//We need to not continue if there is audio playing, so we just pause here.
-			else if (_playingSound)
-			{
-				if (!GetComponent<AudioSource>().isPlaying)
-				{
-					//We finish it here
-					E.Get().CurrTrial.Progress();
-				}
-			}
-			else
-			{	
-				//This section rotates the camera (potentiall up 15 degrees), basically deprecated code.
-				if (!_reset)
-				{
-					Cam.transform.Rotate (0, 0, 0);
-					_reset = true;
-					TrialProgress.GetCurrTrial().ResetTime();
-				}
-				
-				//Move the character.
-				ComputeMovement();
-				E.LogData(
-					TrialProgress.GetCurrTrial().TrialProgress, 
-					TrialProgress.GetCurrTrial().GetRunningTime(),
-					transform
-				);
-			}
-			_currDelay += Time.deltaTime;
-		}
-		
-
-	}
+            }
+            else
+            {
+                var p = E.Get().CurrTrial.Value.CharacterStartPos;
+                if (maze)
+                    p = new List<float>() { pickX, pickY };
+                transform.position = new Vector3(p[0], 0.5f, p[1]);
+                var camPos = Cam.transform.position;
+                camPos.y = DS.GetData().CharacterData.Height;
+                Cam.transform.position = camPos;
+            }
+        }
+
+
+        //This is the collision system.
+        private void OnTriggerEnter(Collider other)
+        {
+            if (!other.gameObject.CompareTag("Pickup")) return;
+
+
+            GetComponent<AudioSource>().PlayOneShot(other.gameObject.GetComponent<PickupSound>().Sound, 10);
+            Destroy(other.gameObject);
+
+            DS.GetData().BlockList[TrialProgress.GetCurrTrial().BlockID].NumCollectedThisBlock++;
```

Comment:
Consider moving this into the TrialProgress object. Should limit Data object storing global state.

Reply:
Hi @hongyu-wang, thanks for your feedback! 👍 

You make a good point about reducing the global state size. However, it seems as though the `TrialProgress` object persists throughout the entire experiment (kept alive via the Loader), so in a sense it's also storing global state - and I was looking for something that would keep track on a per block basis. Rather than reset the counter after each block, I just tacked it on to the block data object :S.

Reply:
I guess responsibility of Data object should be maintaining persistent state as well as initial configuration while TrialProgress maintains the state of the experiment. You can make a list inside TrialProgress to maintain this counter. Data should be entirely intialized by the config file.

Reply:
Ah, I see what you mean - keep all the changing state inside `TrialProgress`. Good idea! I'll make that change now :).